pkg/controller/factory/base_controller.go

```diff
@@ -91,8 +128,25 @@ func (c *baseController) Run(ctx context.Context, workers int) {
 			// If caches did not sync after 10 minutes, it has taken oddly long and
 			// we should provide feedback. Since the control loops will never start,
 			// it is safer to exit with a good message than to continue with a dead loop.
-			// TODO: Consider making this behavior configurable.
-			klog.Exit(err)
+			c.exitOnError(err)
+		}
+	}
+
+	// for preconditions, use the cache sync timeout minus the time we were waiting for caches for sync.
+	preconditionTimeout := c.cacheSyncTimeout - time.Now().Sub(cacheSyncStartTime)
+
+	preconditionCtx, preconditionCtxCancel := context.WithTimeout(ctx, preconditionTimeout)
```

Comment:
you like long var names

Reply:
yeah, they are cheap and describe what is this context for :)